MaxText/train.py

```diff
@@ -211,7 +211,15 @@ def loss_fn(model, config, data, dropout_rng, params, is_train=True):
     for k, v in data.items():
       data[k] = v[:config.global_batch_size_to_train_on,:]
 
-  logits, intermediate_outputs = model.apply({'params': params},
+  if config.quantization=="fp8":
```

Comment:
I don't understand this forking. It is because of the difference between passing in `params` and {'params' : params}? Let's find a solution without that forking

Reply:
cc @kaixih, do you know if there is a way around this?

Reply:
@abhinavgoel95 I'll also poke around with this a bit more today. I think there should be a way to just make this consistent throughout maxtext such that we just pass `params` around directly (so it'll always be {'params': params} by default). But that might take a little bit of extra work in this PR. I'll report back

Reply:
The current behavior of TrainState is such that the pytree structure of optimizer states depends on what the user provides:

1. If the input is {'params': {'w': xxx, 'b': xxx}, 'other_params': yyy}, the optimizer state will be {'params': {'w': xxx, 'b': xxx}, 'other_params': yyy}.
2. If the input is {'w': xxx, 'b': xxx}, the optimizer state will be {'w': xxx, 'b': xxx}.
3. If the input is {'params': {'w': xxx, 'b': xxx}, 'other_params': yyy, 'overwrite_with_gradient': zzz}, the optimizer state will be {'w': xxx, 'b': xxx}.

The behavior of (1) and (2) represents the optimizer states when we pass `model_var` and `model_var['params']` into TrainState. However, (3) passes the full model variables while the optimizer state only focuses on 'params' and disregards 'other_params', which might need some improvement if we do have any real-world use case that we need optimizer states for non 'params' variables.

Back to our scenario here, we are employing method (2) by originally passing `model_var['params']`. The proposed change entails utilizing (3) for fp8 configuration by passing the entire `model_vars`.

Considering this, if there are no real instances where `other_params` necessitates optimizer states, I suggest modifying the TrainState behavior in Flax for (1) as follows:

(1) {'params': {'w': xxx, 'b': xxx}, 'other_params': yyy} -> opt {'w': xxx, 'b': xxx}

With this adjustment, the TrainState behavior remains consistent: it will generate optimizer states only for 'params' if they exist (cases 1 and 3); otherwise, it will generate states for all items (case 2). Consequently, there's no need for branching here, and we can simply pass in the entire `model_vars`:
```python
init_training_state(model.apply, model_vars, tx)

```

Reply:
@kaixih Thank you for the thorough explanation! The behavior in (3) is strange to me - I don't understand why it differs from (1). I can raise this internally with the Flax team. That said, I think I might have a workaround for us here: 

@abhinavgoel95 For this PR, I did a little bit of debugging of MaxText (on a TPU, but I think this should generalize). Within `init_initial_state`, `model_vars` looks like the following: 

`{'params': {'token_embedder': {...}, 'decoder': {...}}}`

So we can just pass it directly to `init_training_state` directly. And then within `loss_fn` in `train.py`, we can just pass `params` along directly. With just those changes, I was able to get training running. And I believe it should generalize to your changes (I'll try to verify that today). 

That said, I think MaxText has been a little inconsistent with this params versus params[params], so I think we might get some test errors, so it might be take a bit of work to get this done properly.

Reply:
One other thing that would be good to have in this PR is some kind of additional smoke test for the fp8 case. Along with the `params` testing, that is increasing the scope of the PR, so it might be good to separate the two out into separate PRs (once we verify that the change I proposed actually helps) - 1 to switch just using params everywhere instead of params[parmas]) and then the fp8 work + test. If you have bandwidth/want to do both, great! If not, I'm happy to do the first PR and then you can incorporate the changes here.

Reply:
@abhinavgoel95 I ran the code on H100 with FP8 quantization, but faced following issue.

```
flax.errors.ScopeCollectionNotFound: Tried to access "input_amax_history" from collection "_overwrite_with_gradient" in "/decoder/remat(scan(layers))/self_attention/query/Fp8DotGeneralOp_0" but the collection is empty. (https://flax.readthedocs.io/en/latest/api_reference/flax.errors.html#flax.errors.ScopeCollectionNotFound)
```

Is this error due to the discussion above?

Reply:
@hxssgaa That looks to be related to some errors related to scan layers. That will be fixed before this PR is merged. You can try setting `scan_layers` to False in the meantime

Reply:
Should be fixed now. Both with and without scan works fine.